p2p/src/peer.rs

```diff
@@ -176,70 +243,85 @@ where
         }
     }
 
-    async fn handshake(&mut self) -> Result<(), Error> {
+    async fn handshake(&mut self) -> Result<&Self, Error> {
```

Comment:
This is a very strange signature.

Reply:
It's there to make the methods chainable. It was an attempt to re-factor the `handshake` method and get rid of the odd while loop with `continue`.